File: scripts/fetch-images.py
#!/usr/bin/env python3
"""Download card images from berserk-nxt repository."""

import json
import urllib.error
import urllib.request
from concurrent.futures import ThreadPoolExecutor, as_completed
from pathlib import Path

ROOT = Path(__file__).resolve().parent.parent
CARDS_PATH = ROOT / "public" / "data" / "cards.json"
OUT_DIR = ROOT / "public" / "cards"
BASE_URL = (
    "https://raw.githubusercontent.com/SkAZi/berserk-nxt/main/resources/cards"
)


def download_card(set_id: int, number: int) -> bool:
    dest = OUT_DIR / str(set_id) / f"{number}.webp"
    if dest.exists():
        return True

    url = f"{BASE_URL}/{set_id}/{number}.webp"
    try:
        with urllib.request.urlopen(url) as response:
            dest.parent.mkdir(parents=True, exist_ok=True)
            dest.write_bytes(response.read())
        return True
    except urllib.error.HTTPError:
        return False


def main() -> None:
    cards = json.loads(CARDS_PATH.read_text(encoding="utf-8"))
    targets = [
        (card["setId"], card["number"])
        for card in cards
        if not card.get("nonStandardLayout")
    ]

    downloaded = 0
    missing = 0

    with ThreadPoolExecutor(max_workers=8) as executor:
        futures = {
            executor.submit(download_card, set_id, number): (set_id, number)
            for set_id, number in targets
        }
        for future in as_completed(futures):
            if future.result():
                downloaded += 1
            else:
                missing += 1

    print(
        f"Images: {downloaded} downloaded, {missing} missing ({len(targets)} requested)"
    )


if __name__ == "__main__":
    main()

File: scripts/fetch-images.mjs
import { createWriteStream } from 'node:fs'
import { mkdir, readFile } from 'node:fs/promises'
import { join } from 'node:path'
import { pipeline } from 'node:stream/promises'
import { spawn } from 'node:child_process'

const ROOT = process.cwd()
const CARDS_PATH = join(ROOT, 'public', 'data', 'cards.json')

async function runOfficialMapping() {
  await new Promise((resolve, reject) => {
    const child = spawn('python', ['scripts/fetch-images-official.py'], {
      cwd: ROOT,
      stdio: 'inherit',
      shell: process.platform === 'win32',
    })
    child.on('exit', (code) => {
      if (code === 0) resolve(undefined)
      else reject(new Error(`fetch-images-official.py exited with ${code}`))
    })
  })
}

async function downloadLocalImages() {
  const cards = JSON.parse(await readFile(CARDS_PATH, 'utf-8'))
  const targets = cards.filter((card) => card.imageUrl)

  let downloaded = 0
  let missing = 0

  for (const card of targets) {
    const dest = join(
      ROOT,
      'public',
      'cards',
      String(card.setId),
      `${card.number}.webp`,
    )

    try {
      const response = await fetch(card.imageUrl)
      if (!response.ok) {
        missing += 1
        continue
      }

      await mkdir(join(dest, '..'), { recursive: true })
      await pipeline(response.body, createWriteStream(dest))
      downloaded += 1
    } catch {
      missing += 1
    }
  }

  console.log(`Local cache: ${downloaded} saved, ${missing} failed`)
}

async function runFillMissing() {
  await new Promise((resolve, reject) => {
    const child = spawn('python', ['scripts/fill-missing-images.py'], {
      cwd: ROOT,
      stdio: 'inherit',
      shell: process.platform === 'win32',
    })
    child.on('exit', (code) => {
      if (code === 0) resolve(undefined)
      else reject(new Error(`fill-missing-images.py exited with ${code}`))
    })
  })
}

async function main() {
  console.log('Step 1: map official berserk.ru image URLs...')
  await runOfficialMapping()

  console.log('Step 2: fill remaining gaps from URL patterns...')
  await runFillMissing()

  if (process.argv.includes('--local')) {
    console.log('Step 3: cache images locally...')
    await downloadLocalImages()
  } else {
    console.log('Skipping local download. Cards will use remote imageUrl.')
  }
}

main().catch((error) => {
  console.error(error)
  process.exit(1)
})
